usb: usbd: msc: improve the performance of msc

MarkWangChinese opened the pull request:
increase the buf size for every transfer giving to controller, reduce the memory copy.

MarkWangChinese commented:
### test with mcxn947 HS USB controller (EHCI)
- with this patch
build cmd: ` west build -b mcx_n9xx_evk/mcxn947/cpu0 .\samples\subsys\usb\mass -T sample.usb_device_next.mass_sdhc_fatfs`  and set `CONFIG_SPEED_OPTIMIZATIONS=y` and `CONFIG_USBD_MSC_DATA_BUF_SIZE=32768`
<img width="835" height="277" alt="image" src="https://github.com/user-attachments/assets/5ceca80b-1c34-4b34-a1f8-772f79acaacc" />

- without this patch
build cmd: ` west build -b mcx_n9xx_evk/mcxn947/cpu0 .\samples\subsys\usb\mass -T sample.usb_device_next.mass_sdhc_fatfs`  and set `CONFIG_SPEED_OPTIMIZATIONS=y` and `CONFIG_USBD_MSC_SCSI_BUFFER_SIZE=32768`
<img width="772" height="246" alt="image" src="https://github.com/user-attachments/assets/46932c30-9243-4e81-9cd9-ab95bfe65d2e" />

tmon-nordic commented on subsys/usb/device_next/class/usbd_msc.c:
```diff
@@ -61,10 +61,18 @@ struct CSW {
+		    MSC_NUM_INSTANCES * 2, USBD_MAX_BULK_MPS,
+		    sizeof(struct udc_buf_info), NULL);
+
+UDC_BUF_POOL_DEFINE(msc_ep_pool_data,
+		    MSC_NUM_INSTANCES, MSC_BUF_SIZE,
```
I was thinking about the "combine `msc_ep_pool` and `msc_ep_pool_data` as one variable-size pool?" when `CONFIG_USBD_MSC_DATA_BUF_SIZE` is greater than `USBD_MAX_BULK_MPS` (otherwise the single pool would be fixed size pool). I think the minimum pool size would then be `USBD_MAX_BULK_MPS + CONFIG_USBD_MSC_DATA_BUF_SIZE`. It is up to the code to not try to allocate more than `CONFIG_USBD_MSC_DATA_BUF_SIZE`, but as this pool is only used within this particular class implementation, ensuring this shouldn't be a problem.

MarkWangChinese replied:
Thanks. I add one commit to do it, please help to review again. Maybe I do more changes than you thought.

tmon-nordic replied:
I think the two commits should be squashed as first you add some complex handling and then it is removed immediately in next commit.

tmon-nordic commented on subsys/usb/device_next/class/usbd_msc.c:
```diff
@@ -125,14 +136,20 @@ struct msc_bot_ctx {
 	uint32_t transferred_data;
 	size_t scsi_offset;
 	size_t scsi_bytes;
+	bool pending_read;
```
Is this still necessary now that there is just one pool?

MarkWangChinese replied:
Yes, the checking is still needed, but I change to use `scsi_cmd_remaining_data_len(lun)`.
In the original codes logic, `ctx->scsi_bytes == 0` can indicate whether there is still data for bulk in. Since my changes use the net_buf to do the scsi reading, the `ctx->scsi_bytes == 0` is true even there is still data for bulk in.

tmon-nordic commented on subsys/usb/device_next/class/Kconfig.msc:
```diff
@@ -30,6 +30,12 @@ config USBD_MSC_SCSI_BUFFER_SIZE
 
+config USBD_MSC_DATA_BUF_SIZE
+	int "Data Buffer size for msc data transfer"
+	default 512
```
```suggestion
	default 512 if UDC_DRIVER_HAS_HIGH_SPEED_SUPPORT
	default 64
```

would maintain default low memory usage on Full-Speed only drivers.

tmon-nordic commented:
I tried it with `CONFIG_USBD_MSC_SCSI_BUFFER_SIZE` set to 131072 and while it was calling `fill_read_10()` with length 131072, the actual transfers submitted on IN endpoint were just 512. When there is large SCSI buffer, why not use it directly?

The copying between scsi buffer and msc buffer is for sure. `net_buf_alloc_with_data()` can be used to use SCSI buffer directly. The only requirement would be to make the scsi buffer aligned to `UDC_BUF_ALIGN` and be multiple of `UDC_BUF_GRANULARITY`.

tmon-nordic reviewed (changes requested):
I have added necessary disk size constraint checks and cleaned up the code a bit. Please pick up the modified commit from https://github.com/tmon-nordic/zephyr/tree/msc-zero-copy (force push it to your PR branch).

MarkWangChinese committed:
usb: device_next: msc: do not copy SCSI data

Support only power-of-two disk sector sizes to enable significant
throughput improvements:
  * SCSI data zero-copy
  * allow queuing multi-packet transfers

Previously large SCSI buffers did not improve performance. With this
change, larger SCSI buffer allows scheduling bigger USB transfers which
reduces overhead.

Co-authored-by: Tomasz Moń <tomasz.mon@nordicsemi.no>
Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>
Signed-off-by: Mark Wang <yichang.wang@nxp.com>

jfischer-no reviewed (commented):
LGTM, just two comments.
